spec/ics-007-tendermint-client/README.md

```diff
@@ -378,6 +385,10 @@ function verifyPacketAcknowledgement(
     assert(clientState.latestHeight >= height)
     // check that the client is unfrozen or frozen at a higher height
     assert(clientState.frozenHeight === null || clientState.frozenHeight > height)
+    // fetch the processed time
+    processedTime = get("clients/{identifier}/processedTimes/{height}")
```

Comment:
I am puzzled here. Let's say packet p with delay d was sent from chain A to chain B at height h. Packet is correctly received at chain B if there is a header for height h+1 installed at time t and packet is received not before time t+d. Verifying packet acknowledgment for p is happening on the chain A, and if we want to verify that packet was indeed not processed before t+d, we will need to have access to the consensusState at chain B at the header reception time and processingTime of the packet p at chain B. So both these information need to be carried to the chain A with the corresponding proofs so chain A can verify that indeed packet was correctly processed on B. Does this make sense?

Reply:
Hmm. Maybe I'm misunderstanding the intended model here. I thought we were trying to ensure that after a given header was submitted to a chain - call it *C* - a period of time specified per-packet must pass before any proof verification involving that packet or related data (such as the acknowledgement) which uses that header is allowed to complete, so that there would be time for misbehaviour to be submitted to the chain. I don't see what relevance the timestamp on the other chain would have to this since it's the light client on *this* chain which we want time to submit misbehaviour to - also, the timestamp on the other chain is not a local clock, so it's not clear how we would compare it. Am I misunderstanding something fundamental here?

Reply:
So high level invariant we want is the following: packet verification on the reception side should fail if delayPeriod hasn't passed since the corresponding header verification/update. This allows correct relayer to detect faulty header and shutdown light client. Now, as packet reception is a precondition for valid acknowledgment processing on the sender side, acknowledgment verification depends on it. In the model where we trust state transitions to be valid if we see valid light client proof, nothing would need to be changed on the sender side (ack/timeout verification). Note that in the model without delay this is the case as we only verify light client proofs, and not if the corresponding transition was executed (if I am not wrong). In this concrete case we have a way to verify if time has passed, so we could indeed validate this on the sender side, but I think we will need more data in ack/timeout packets to be able to do this. The question is if that is needed.

Reply:
> Let's say packet p with delay d was sent from chain A to chain B at height h. Packet is correctly received at chain B if there is a header for height h+1 installed at time t and packet is received not before time t+d. 

Correct, I think we are all agreed on this.

>  Now, as packet reception is a precondition for valid acknowledgment processing on the sender side, acknowledgment verification depends on it. In the model where we trust state transitions to be valid if we see valid light client proof, nothing would need to be changed on the sender side (ack/timeout verification). Note that in the model without delay this is the case as we only verify light client proofs, and not if the corresponding transition was executed (if I am not wrong). In this concrete case we have a way to verify if time has passed, so we could indeed validate this on the sender side, but I think we will need more data in ack/timeout packets to be able to do this. 

So from my understanding you're talking about the acknowledgement logic verifying that the packet delay period had indeed passed on the counterparty before packet receipt was processed. ie ChainA verifies that chainB waited for delay period before receiving, and only then will successfully ACK. (and similarly with timeouts)

I'm not sure this is what we want to do, compared with Chris's comment which is to enforce the packet-specified delay period for any packet-related verification that occurs on our own chain. And not check that packet-delay is correctly enforced on the counterparty chain.

Going back to your example, a user sends a packet p with delay period d from chainA to chainB.

Assuming a correct implementation on chainB, the consensus state used to verify chainA's packet commitment must at least exist for `d` time before a relayer is allowed to use it to receive the packet on chainB.
**Note that this feature protects the recipient on chainB from potential malicious behaviour of chainA.**

Now there are only 2 instances in my mind that a chainB will receive a packet before the user-specified delay period.

**1. Incorrect implementation:**

In this case, chainB verifies the packet without enforcing the delay period. One could imagine an alternate IBC implementation that simply ignored the packet's delay period.
Note from above that this is a choice the chain is allowed to make that simply affects its own users. ie the recipient on chainB is vulnerable to misbehaviour on chainA because chainB chose not to enforce delay period.
In this case, I don't believe chainA should care whether chainB follows delay period or not as it does not affect chainA.
Why should chainA enforce that chainB protects users against chainA becoming malicious? I believe it is safe for a chain to assume itself to be honest and avoid unnecessary verification work that way.
Furthermore, suppose chainB does ignore the delay period and receives the packet anyway. It's an open question what should be done in light of this.
Should chainA never acknowledge the packet since chainB incorrectly received before delay period? I believe this is what your proposal would do since if we pass the counterparty's consensus state and processing time back to chainA and the delay period check fails, chainB will have received the packet while chainA can never acknowledge that receipt because it will always fail the delay period check.
This can be dangerous for applications. Take transfer for example, if chainB receives the packet before delay period and for whatever reason errors and returns an `Failed` acknowledgement. chainA will never be able to process that acknowledgement and refund the sender.

I believe it is safe and simpler for chainA to just not check whether chainB enforces the delay period during packet receipt or not. What is important is that each chain enforces the delay period on its own end to protect its own users from misbehaviour by counterparties. It is unnecessary for chainA to check that counterparties are protecting their users from potential misbehaviour by chainA.

**2. Fork**

The other case where chainA may try to ack a packet receipt is if there is a malicious fork on chainB such that the malicous fork has received the packet before delay period. A relayer might then submit a proof against the malicous fork to chainA and try to get the ack processed.

However, this case is already handled by the current spec and implementation since the packet-delay period specified when the packet is sent will apply to **all packet verification logic**, including acks and timeouts. Thus the consensus state on chainA must exist for at least `d` time before a relayer can use it to verify an ACK on chainB.

This offers the same level of misbehaviour protection for acks and timeouts as well, since a relayer will have at least `d` time to submit misbehaviour that could happen at any stage in the packet lifecycle.

Reply:
> So high level invariant we want is the following: packet verification on the reception side should fail if delayPeriod hasn't passed since the corresponding header verification/update.

I believe this invariant should only be enforced on the receiver side and not the sender side for the reasons stated above.

I think the invariant we want to enforce is actually this: Each chain enforces a delay period before accepting any packet-level verification (receipts/acks/timeouts) on its own chain to allow time for misbehaviour to be submitted against a counterparty.

Reply:
Thanks a lot @AdityaSripal for clarifying all this. I misunderstood the ack/timeout logic. I thought it's related to packet reception and as you nicely explained it ensures the same guarantee for ack/timeout data. I think that I agree with your proposal regarding 1, but we can synchronously confirm it during the IBC call.

Reply:
Sounds like we're in agreement here (thanks @AdityaSripal for the clearer explanation) but let's confirm on the call.